test/e2e/kubelet_stats.go

```diff
@@ -287,6 +356,102 @@ func formatResourceUsageStats(nodeName string, containerStats map[string]*contai
 	return fmt.Sprintf("Resource usage on node %q:\n%s", nodeName, buf.String())
 }
 
+type int64arr []int64
+
+func (a int64arr) Len() int           { return len(a) }
+func (a int64arr) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a int64arr) Less(i, j int) bool { return a[i] < a[j] }
+
+func computePercentiles(timeSeries map[time.Time]resourceUsagePerContainer, percentilesToCompute []int) map[int]resourceUsagePerContainer {
+	cpuData := make(map[string][]float64)
+	memUseData := make(map[string][]int64)
+	memWorkSetData := make(map[string][]int64)
+	for _, singleStatistic := range timeSeries {
+		for name, data := range singleStatistic {
+			cpuData[name] = append(cpuData[name], data.CPUUsageInCores)
+			memUseData[name] = append(memUseData[name], data.MemoryUsageInBytes)
+			memWorkSetData[name] = append(memWorkSetData[name], data.MemoryWorkingSetInBytes)
+		}
+	}
+	for _, v := range cpuData {
+		sort.Float64s(v)
+	}
+	for _, v := range memUseData {
+		sort.Sort(int64arr(v))
+	}
+	for _, v := range memWorkSetData {
+		sort.Sort(int64arr(v))
+	}
+
+	result := make(map[int]resourceUsagePerContainer)
+	for _, perc := range percentilesToCompute {
+		data := make(resourceUsagePerContainer)
+		for k, v := range cpuData {
+			data[k] = &containerResourceUsage{
+				Name:                    k,
+				CPUUsageInCores:         v[int(math.Ceil(float64(len(v)*perc)/100))-1],
+				MemoryUsageInBytes:      memUseData[k][int(math.Ceil(float64(len(v)*perc)/100))-1],
+				MemoryWorkingSetInBytes: memWorkSetData[k][int(math.Ceil(float64(len(v)*perc)/100))-1],
+			}
+		}
+		result[perc] = data
+	}
+	return result
+}
+
+type containerResourceGatherer struct {
+	usageTimeseries map[time.Time]resourceUsagePerContainer
+	stopCh          chan struct{}
+	timer           *time.Ticker
+	wg              sync.WaitGroup
+}
+
+func (g *containerResourceGatherer) startGatheringData(c *client.Client, period time.Duration) {
+	g.usageTimeseries = make(map[time.Time]resourceUsagePerContainer)
+	g.wg.Add(1)
+	g.stopCh = make(chan struct{})
+	g.timer = time.NewTicker(period)
+	go func() error {
+		for {
+			select {
+			case <-g.timer.C:
+				now := time.Now()
+				data, err := getKubeSystemContainersResourceUsage(c)
+				if err != nil {
+					return err
+				}
+				g.usageTimeseries[now] = data
+			case <-g.stopCh:
+				g.wg.Done()
+				return nil
+			}
+		}
+	}()
+}
+
+func (g *containerResourceGatherer) stopAndPrintData(percentiles []int) {
+	close(g.stopCh)
+	g.timer.Stop()
+	g.wg.Wait()
+	stats := computePercentiles(g.usageTimeseries, percentiles)
+	sortedKeys := []string{}
+	for name := range stats[percentiles[0]] {
```

Comment:
This assumes that there is at least one element in the map - what if it is empty?